Include/pythread.h

```diff
@@ -77,15 +79,67 @@ PyAPI_FUNC(int) PyThread_set_stacksize(size_t);
 PyAPI_FUNC(PyObject*) PyThread_GetInfo(void);
 #endif
 
-/* Thread Local Storage (TLS) API */
-PyAPI_FUNC(int) PyThread_create_key(void);
-PyAPI_FUNC(void) PyThread_delete_key(int);
-PyAPI_FUNC(int) PyThread_set_key_value(int, void *);
-PyAPI_FUNC(void *) PyThread_get_key_value(int);
-PyAPI_FUNC(void) PyThread_delete_key_value(int key);
+/* Thread Local Storage (TLS) API
+   TLS API is DEPRECATED.  Use Thread Specific Storage API.
+*/
+PyAPI_FUNC(int) PyThread_create_key(void) Py_DEPRECATED(3.7);
+PyAPI_FUNC(void) PyThread_delete_key(int key) Py_DEPRECATED(3.7);
+PyAPI_FUNC(int) PyThread_set_key_value(int key, void *value) Py_DEPRECATED(3.7);
+PyAPI_FUNC(void *) PyThread_get_key_value(int key) Py_DEPRECATED(3.7);
+PyAPI_FUNC(void) PyThread_delete_key_value(int key) Py_DEPRECATED(3.7);
+
+/* Cleanup after a fork */
+PyAPI_FUNC(void) PyThread_ReInitTLS(void) Py_DEPRECATED(3.7);
+
+/* Thread Specific Storage (TSS) API
+
+   POSIX hasn't defined that pthread_key_t is compatible with int
+   (for details, see PEP 539).  Therefore, TSS API uses opaque type to cover
+   the key details.
+*/
+
+#if defined(_POSIX_THREADS)
+#   define NATIVE_TLS_KEY_T     pthread_key_t
+#elif defined(NT_THREADS)
+#   include <windows.h>
+#   define NATIVE_TLS_KEY_T     DWORD
+#else  /* For the platform that has not supplied native TLS */
+#   define NATIVE_TLS_KEY_T     int
+#endif
+
+/* Py_tss_t is opaque type and you *must not* directly read and write.
+   When you'd check whether the key is created, use PyThread_tss_is_created.
+*/
+typedef struct {
+    bool _is_initialized;
+    NATIVE_TLS_KEY_T _key;
+} Py_tss_t;
+
+#undef NATIVE_TLS_KEY_T
+
+static inline bool
+PyThread_tss_is_created(Py_tss_t key)
+{
+    return key._is_initialized;
+}
```

Comment:
With a `PyThread_tss_alloc`, would the need for `Py_tss_NEEDS_INIT` to be public go away?  After all, if one used `PyThread_tss_alloc` to allocate memory for a struct, what else would they even initialize it with?  It would remove an extra step to just do that as part of `PyThread_tss_alloc` and remove that implementation detail.  

It should just be clear that just because a TSS key has been *allocated* doesn't mean it's been *created*.  Also, `PyThread_tss_is_created` is still available to test whether or not a key has been created.